packages/sdk/src/fee/FeeEstimator.ts

```diff
@@ -0,0 +1,164 @@
+import { Psbt } from "bitcoinjs-lib";
+import { Buffer } from "buffer";
+
+import type { AddressFormat } from "../addresses/types";
+import type { Network } from "../config/types";
+import { MAXIMUM_FEE } from "../constants";
+import { OrditSDKError } from "../errors";
+import { getNetwork, getScriptType } from "../utils";
+import type { FeeEstimatorOptions } from "./types";
+
+function getBaseSizeByType(type: AddressFormat) {
+  // Refer to BIP-0141 - https://github.com/bitcoin/bips/blob/master/bip-0141.mediawiki
+  // Calculator - https://bitcoinops.org/en/tools/calc-size/
+  switch (type) {
+    case "taproot":
+      return { input: 42, output: 43, txHeader: 10.5, witness: 66 }; // witness size is different for non-default sigHash
+
+    case "segwit":
+      return { input: 41, output: 31, txHeader: 10.5, witness: 105 };
+
+    case "p2sh-p2wpkh":
+      return { input: 64, output: 32, txHeader: 10, witness: 105 };
+
+    case "legacy":
+      return { input: 148, output: 34, txHeader: 10, witness: 0 };
+
+    default:
+      throw new OrditSDKError("Invalid type");
+  }
+}
+
+class FeeEstimator {
+  protected fee = 0;
+
+  protected feeRate: number;
+
+  protected network: Network;
+
+  protected psbt: Psbt;
+
+  protected witness?: Buffer[] = [];
+
+  protected virtualSize = 0;
+
+  protected weight = 0;
+
+  constructor({ feeRate, network, psbt, witness }: FeeEstimatorOptions) {
+    if (feeRate < 0 || !Number.isSafeInteger(feeRate)) {
+      throw new OrditSDKError("Invalid feeRate");
+    }
+
+    this.feeRate = feeRate;
+    this.network = network;
+    this.witness = witness || [];
+    this.psbt = psbt || new Psbt({ network: getNetwork(this.network) });
+  }
+
+  get data() {
+    return {
+      fee: this.fee,
+      virtualSize: this.virtualSize,
+      weight: this.weight,
+    };
+  }
+
+  calculateNetworkFee(): number {
+    this.fee = this.calculateVirtualSize() * this.feeRate;
+    // Prevents catastrophic calculations from happening
+    if (this.fee > MAXIMUM_FEE) {
+      throw new OrditSDKError("Error while calculating fees");
+    }
+
+    return this.fee;
+  }
+
+  private getInputAndOutputScriptTypes() {
+    const { inputs } = this.psbt.data;
+    const outputs = this.psbt.txOutputs;
+
+    if (inputs.length === 0) {
+      throw new OrditSDKError("PSBT must have at least one input");
+    }
+
+    if (outputs.length === 0) {
+      throw new OrditSDKError("PSBT must have at least one output");
+    }
+
+    return {
+      inputTypes: inputs.map((input) => {
+        // P2PKH (Legacy) is not supported as there is no witness script.
```

Comment:
I was looking at bitcoinjs-lib transaction tests and P2PKH uses `nonWitnessUtxo` - do we intend to support this?

Reply:
I was referring to the hacky way that I used below:

https://github.com/sadoprotocol/ordit-sdk/blob/089683ebd37f74e86ac596f924ba444ee796bccd/packages/sdk/src/transactions/psbt.ts#L149-L153

Script type and sats value was needed to process further so `witnessUtxo` has been used here to pass that. I would suggest to remove the comment as it might mislead whoever reads the code and add a commend in the original fn above explaining why we are passing data in `witnessUtxo`.

Reply:
I've removed the comment.

I'm not sure what you mean by process further, do you mean for seller psbts, or for segments in the code where we call `processInput`?